ion-schema/src/nfa.rs

```diff
@@ -0,0 +1,258 @@
+                *source_state_id,
+                transitions,
+                can_exit,
+                *visits,
+                event.to_owned(),
+                type_store,
+                &mut new_states,
+            );
+        }
+        new_states
+    }
+
+    fn evaluate_transitions(
+        &self,
+        source_state_id: usize,
+        transitions: &HashSet<Transition>,
+        can_exit: bool,
+        visits: usize,
+        event: Event,
+        type_store: &TypeStore,
+        new_states: &mut HashMap<usize, usize>,
+    ) {
+        let mut destination_states_for_optional_state: HashSet<Transition> = HashSet::new();
+        for transition in transitions {
+            let destination_state_id = transition.destination;
+            let destination_state = &self.nfa.states[destination_state_id];
+
+            if destination_state_id == source_state_id {
+                if self.nfa.states[destination_state_id].allows_n_visits(visits + 1) {
+                    let can_enter = destination_state.allows_to_enter_for_ion_value(
+                        event.to_owned(),
+                        transition.type_id,
+                        type_store,
+                    );
+                    if can_enter {
+                        new_states.insert(destination_state_id, visits + 1);
+                    }
+                }
+            } else if can_exit {
+                let can_enter = destination_state.allows_to_enter_for_ion_value(
+                    event.to_owned(),
+                    transition.type_id,
+                    type_store,
+                );
+                if can_enter {
+                    new_states.insert(destination_state_id, 1);
+                }
+            }
+
+            // for optional state add transitions to next states skipping the optional state
+            if transition.to_optional_state {
+                let mut transitions = self
+                    .nfa
+                    .transitions
+                    .get(&destination_state_id)
+                    .unwrap()
+                    .to_owned();
+
+                // skip the optional state itself
+                transitions.remove(transition);
+
+                destination_states_for_optional_state.extend(transitions);
+
+                self.evaluate_transitions(
+                    destination_state_id,
+                    &destination_states_for_optional_state,
+                    can_exit,
+                    visits,
+                    event.to_owned(),
+                    type_store,
+                    new_states,
+                );
+            }
+        }
+    }
+}
+
+/// Represents a builder for constructing NFA which will be used by `ordered_elements` constraint
+pub struct NfaBuilder {
+    nfa: Nfa,
+}
+
+impl NfaBuilder {
+    pub fn new() -> NfaBuilder {
+        NfaBuilder {
+            nfa: Nfa::default(),
+        }
+    }
+
+    pub fn build(self) -> Nfa {
+        self.nfa
+    }
```

Comment:
suggestion—this should check some of the invariants of the NFA graph.
- do all of the transitions start at a valid state id?
- do all of the transitions end at a valid state id?